### PR TITLE
UCT/IB: Allow QP for pkey with limited membership

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Allow QP initialization for pkey with limited membership.

## Why ?

To allow communications between peers with limited and full membership (L<->F, F<->L, F<->F).

## How ?

1. Allow using PKEY with limited membership if it was requested thru environment variable.
2. Pack PKEY value (if it is not default - 0xffff) to the IB address and update UCT reachability function.

### CODE REVIEW

**Comment by dmitrygx** [comment]
waiting for the testing results on multi-PKEY configuration

File: src/uct/ib/base/ib_iface.c [hunk 1]
```diff
@@ -385,25 +403,27 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
 {
     const void *ptr = ib_addr + 1;
 
-    *lid       = 0;
-    *gid_index = UCT_IB_ADDRESS_INVALID_GID_INDEX;
-    *path_mtu  = UCT_IB_ADDRESS_INVALID_PATH_MTU;
+    *lid        = 0;
+    *gid_index  = UCT_IB_ADDRESS_INVALID_GID_INDEX;
+    *path_mtu   = UCT_IB_ADDRESS_INVALID_PATH_MTU;
+    *pkey_value = 0;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:426]: i'd set it to `UCT_IB_PKEY_VALUE_DEFAULT` here to avoid checking pkey_value for `0` in other places

File: src/uct/ib/base/ib_iface.h [hunk 2]
```diff
@@ -69,7 +71,8 @@ enum {
     UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID  = UCS_BIT(1),
     UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX = UCS_BIT(2),
     UCT_IB_ADDRESS_PACK_FLAG_PATH_MTU      = UCS_BIT(3),
-    UCT_IB_ADDRESS_PACK_FLAG_GID_INDEX     = UCS_BIT(4)
+    UCT_IB_ADDRESS_PACK_FLAG_GID_INDEX     = UCS_BIT(4),
+    UCT_IB_ADDRESS_PACK_FLAG_PKEY_VALUE    = UCS_BIT(5),
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.h:75]: no need for comma in the end

File: src/uct/ib/base/ib_iface.h [hunk 3]
```diff
@@ -420,10 +426,12 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
  *                         @ref UCT_IB_ADDRESS_INVALID_GID_INDEX if not present.
  * @param [out] path_mtu   Filled with address path MTU, or
  *                         @ref UCT_IB_ADDRESS_INVALID_PATH_MTU if not present.
+ * @param [out] pkey_value Filled with PKEY value, or 0 if the default (0xffff)
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.h:429]: better fill with default value by default

File: test/gtest/uct/ib/test_ib.cc [hunk 4]
```diff
@@ -9,11 +9,15 @@
 test_uct_ib::test_uct_ib() : m_e1(NULL), m_e2(NULL) { }
 
 void test_uct_ib::create_connected_entities() {
-    m_e1 = uct_test::create_entity(0);
-    m_e2 = uct_test::create_entity(0);
+    if (m_e1 == NULL) {
```
Comment (brminich) [inline test/gtest/uct/ib/test_ib.cc:12]: why need to check for NULL?

File: test/gtest/uct/ib/test_ib_pkey.cc [hunk 5]
```diff
@@ -66,36 +78,156 @@ class test_uct_ib_pkey : public test_uct_ib_with_specific_port {
...
+                              &pkey_value1);
+        if (m_pkey_value[0] != UCT_IB_PKEY_VALUE_DEFAULT) {
+            EXPECT_EQ(pkey_value1, m_pkey_value[0]);
+        } else {
+            EXPECT_EQ(0, pkey_value1);
+            pkey_value1 = UCT_IB_PKEY_VALUE_DEFAULT;
+        }
+
+        /* pack-unpack the second IB iface address */
+        uint16_t lid2;
```
Comment (brminich) [inline test/gtest/uct/ib/test_ib_pkey.cc:200]: plz define some member func for common code

Comment (dmitrygx) [reply test/gtest/uct/ib/test_ib_pkey.cc]: done

**Comment by yosefe** [comment]
failure seems relevant:
```
2020-05-18T11:37:26.5259789Z [ RUN      ] ud_mlx5/test_ud_ds.if_addr/0 <ud_mlx5/mlx5_0:1>
2020-05-18T11:37:26.5526836Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/uct/ib/test_ud_ds.cc:85: Failure
2020-05-18T11:37:26.5527740Z Value of: pkey_value1
2020-05-18T11:37:26.5528343Z   Actual: 65535
2020-05-18T11:37:26.5528830Z Expected: 0
2020-05-18T11:37:26.5529711Z /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/uct/ib/test_ud_ds.cc:86: Failure
2020-05-18T11:37:26.5530256Z Value of: pkey_value2
2020-05-18T11:37:26.5530578Z   Actual: 65535
2020-05-18T11:37:26.5530860Z Expected: 0
2020-05-18T11:37:26.5575783Z [  FAILED  ] ud_mlx5/test_ud_ds.if_addr/0, where GetParam() = ud_mlx5/mlx5_0:1 (32 ms)
2020-05-18T11:37:26.5577614Z [----------] 1 test from ud_mlx5/test_ud_ds (32 ms total)
```

**Comment by dmitrygx** [comment]
> failure seems relevant:

@yosefe thanks for pointing me out, fixed

**Comment by dmitrygx** [comment]
@brminich @hoopoepg @evgeny-leksikov could you review pls?
@evgeny-leksikov this is recently touched by your code - IB pack/unpack

File: src/uct/ib/base/ib_iface.c [hunk 6]
```diff
@@ -475,6 +501,10 @@ const char *uct_ib_address_str(const uct_ib_address_t *ib_addr, char *buf,
         snprintf(p, endp - p, "mtu %zu ", uct_ib_mtu_value(mtu));
     }
 
+    if (pkey_value != 0) {
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:504]: when can it be 0?

File: src/uct/ib/base/ib_iface.c [hunk 7]
```diff
@@ -554,8 +584,22 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface,
     uint8_t gid_index;
     uint16_t lid;
     enum ibv_mtu mtu;
+    uint16_t pkey_value;
+
+    uct_ib_address_unpack(ib_addr, &lid, &gid, &gid_index, &mtu, &pkey_value);
 
-    uct_ib_address_unpack(ib_addr, &lid, &gid, &gid_index, &mtu);
+    if (pkey_value == 0) {
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:591]: when can it be 0?

File: test/gtest/uct/ib/test_ib.cc (hunk 4 above)
Comment (brminich) [reply test/gtest/uct/ib/test_ib.cc:12]: then why need to allocate here?

File: src/uct/ib/base/ib_iface.c (hunk 6 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:504]: removed

File: src/uct/ib/base/ib_iface.c (hunk 7 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:591]: removed

File: test/gtest/uct/ib/test_ib.cc (hunk 4 above)
Comment (dmitrygx) [reply test/gtest/uct/ib/test_ib.cc:12]: sorry, I meant that there are two variant:
- where we need to allocate entities here
- where they are allocated from the caller's code

File: src/uct/ib/base/ib_iface.c [hunk 8]
```diff
@@ -475,6 +501,8 @@ const char *uct_ib_address_str(const uct_ib_address_t *ib_addr, char *buf,
         snprintf(p, endp - p, "mtu %zu ", uct_ib_mtu_value(mtu));
     }
 
+    snprintf(p, endp - p, "pkey value 0x%x ", pkey_value);
```
Comment (evgeny-leksikov) [inline src/uct/ib/base/ib_iface.c:504]: need to update `p` after mtu printing

Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:504]: good catch, fixed

File: src/uct/ib/base/ib_iface.c [hunk 9]
```diff
@@ -354,21 +358,35 @@ void uct_ib_address_pack(const uct_ib_address_pack_params_t *params,
         ib_addr->flags |= UCT_IB_ADDRESS_FLAG_GID_INDEX;
         *(uint8_t*)ptr  = params->gid_index;
     }
+
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_PKEY_VALUE) {
+        ucs_assert(params->pkey_value != UCT_IB_ADDRESS_PKEY_VALUE_DEFAULT);
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:363]: maybe just don't pack pkey if it's the default one? this is the logic for subnet prefix for example

File: src/uct/ib/base/ib_iface.c [hunk 10]
```diff
@@ -385,25 +403,27 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
+    params.path_mtu   = UCT_IB_ADDRESS_INVALID_PATH_MTU;
+    params.gid_index  = UCT_IB_ADDRESS_INVALID_GID_INDEX;
+    params.pkey_value = iface->pkey_value;
     uct_ib_address_pack(&params, ib_addr);
 }
 
 void uct_ib_address_unpack(const uct_ib_address_t *ib_addr, uint16_t *lid,
                            union ibv_gid *gid, uint8_t *gid_index,
-                           enum ibv_mtu *path_mtu)
+                           enum ibv_mtu *path_mtu, uint16_t *pkey_value)
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:419]: maybe this function could return *uct_ib_address_pack_params_t ? @evgeny-leksikov  WDYT?
there are already so many values it has to return?
(can be in different PR if it's big change)

File: src/uct/ib/base/ib_iface.c [hunk 11]
```diff
@@ -385,25 +403,27 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
 {
     const void *ptr = ib_addr + 1;
 
-    *lid       = 0;
-    *gid_index = UCT_IB_ADDRESS_INVALID_GID_INDEX;
-    *path_mtu  = UCT_IB_ADDRESS_INVALID_PATH_MTU;
+    *lid        = 0;
+    *gid_index  = UCT_IB_ADDRESS_INVALID_GID_INDEX;
+    *path_mtu   = UCT_IB_ADDRESS_INVALID_PATH_MTU;
+    *pkey_value = UCT_IB_ADDRESS_PKEY_VALUE_DEFAULT;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:426]: UCT_IB_ADDRESS_PKEY_VALUE_DEFAULT -> UCT_IB_ADDRESS_DEFAULT_PKEY_VALUE
also maybe rename pkey_value->pkey in all places (pkey would mean the value and "pkey_index" would be the index when applicable)

File: src/uct/ib/base/ib_iface.c [hunk 12]
```diff
@@ -554,8 +583,18 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface,
...
     uint16_t lid;
     enum ibv_mtu mtu;
+    uint16_t pkey_value;
 
-    uct_ib_address_unpack(ib_addr, &lid, &gid, &gid_index, &mtu);
+    uct_ib_address_unpack(ib_addr, &lid, &gid, &gid_index, &mtu, &pkey_value);
+
+    if (/* at least one PKEY has to be with full membership */
+        (((pkey_value & UCT_IB_PKEY_MEMBERSHIP_MASK) == 0) &&
+         ((iface->pkey_value & UCT_IB_PKEY_MEMBERSHIP_MASK) == 0)) ||
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:592]: shorter : `if ( ! ((pkey_value | iface->pkey_value) & UCT_IB_PKEY_MEMBERSHIP_MASK) ...`

File: src/uct/ib/base/ib_iface.c [hunk 13]
```diff
@@ -695,10 +743,17 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
...
     if (!pkey_found) {
         if (config->pkey_value == UCS_HEXUNITS_AUTO) {
-            ucs_error("There is no valid pkey with full membership on "
-                      UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+            if (!allow_limited && has_pkey_limited) {
+                ucs_debug("there is no valid pkey with full membership on "
+                          UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+                /* retry the search allowing PKEY with limited membership */
+                allow_limited = 1;
+                goto retry;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:751]: maybe simply and use helper function?
e.g 
1. search_for_pkey(full)
2. search_for_pkey(limited)

File: src/uct/ib/base/ib_iface.h [hunk 14]
```diff
@@ -22,6 +22,8 @@
 #define UCT_IB_MAX_ATOMIC_SIZE             sizeof(uint64_t)
 #define UCT_IB_ADDRESS_INVALID_GID_INDEX   UINT8_MAX
 #define UCT_IB_ADDRESS_INVALID_PATH_MTU    0
+/* The default PKEY value is 0xffff */
+#define UCT_IB_ADDRESS_PKEY_VALUE_DEFAULT  UINT16_MAX
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h:26]: just define it to "0xffff"

File: src/uct/ib/base/ib_iface.c (hunk 10 above)
Comment (evgeny-leksikov) [reply src/uct/ib/base/ib_iface.c:419]: 👍 but then need to rename `uct_ib_address_pack_params_t` to smth common and probably unite `UCT_IB_ADDRESS_xx` and `UCT_IB_ADDRESS_PACK_xx` flags. I did not do this change because it'a a big and should be bedefinitely done separately.

**Comment by dmitrygx** [comment]
@evgeny-leksikov could you review pls?

**Comment by dmitrygx** [comment]
bot:pipe:retest

File: src/uct/ib/base/ib_iface.h [hunk 15]
```diff
@@ -97,6 +97,20 @@ typedef struct uct_ib_address_pack_params {
...
+    union ibv_gid                     gid;
+    /* LID address. */
+    uint16_t                          lid;
+    /* path MTU size as defined in enum ibv_mtu. */
+    enum ibv_mtu                      path_mtu;
+    /* GID index. */
+    uint8_t                           gid_index;
+    /* PKEY value. */
+    uint16_t                          pkey;
+} uct_ib_address_unpack_params_t;
```
Comment (evgeny-leksikov) [inline src/uct/ib/base/ib_iface.h:111]: uct_ib_unpacked_address_t

Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.h]: done

File: src/uct/ib/base/ib_iface.c (hunk 9 above)
Comment (yosefe) [reply src/uct/ib/base/ib_iface.c]: for subnet_prefix, the equivalent logic is in uct_ib_address_pack, so lr'ts move the default pkey check to uct_ib_address_pack() as well

File: src/uct/ib/base/ib_iface.h [hunk 16]
```diff
@@ -89,9 +91,26 @@ typedef struct uct_ib_address_pack_params {
     /* GID index,
        must be valid if @ref UCT_IB_ADDRESS_PACK_FLAG_GID_INDEX is set. */
     uint8_t                           gid_index;
+    /* PKEY value,
+       must be valid if @ref UCT_IB_ADDRESS_PACK_FLAG_PKEY is set. */
+    uint16_t                          pkey;
 } uct_ib_address_pack_params_t;
 
 
+typedef struct uct_ib_unpack_address_params {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.h:100]: can we unite uct_ib_address_pack_params and uct_ib_unpack_address_t to same struct?

File: src/uct/ib/base/ib_iface.c [hunk 17]
```diff
@@ -547,19 +574,25 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface,
...
-    uint8_t gid_index;
-    uint16_t lid;
-    enum ibv_mtu mtu;
+    uct_ib_unpack_address_t unpack_addr;
+
+    uct_ib_address_unpack(ib_addr, &unpack_addr);
 
-    uct_ib_address_unpack(ib_addr, &lid, &gid, &gid_index, &mtu);
+    if (/* at least one PKEY has to be with full membership */
+        (((unpack_addr.pkey | iface->pkey) & UCT_IB_PKEY_MEMBERSHIP_MASK) == 0) ||
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:585]: minor: !((unpack_addr.pkey | iface->pkey) & UCT_IB_PKEY_MEMBERSHIP_MASK)

File: src/uct/ib/base/ib_iface.c [hunk 18]
```diff
@@ -547,19 +574,25 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface,
...
+        /* PKEY values have to be equal */
+        ((unpack_addr.pkey & UCT_IB_PKEY_PARTITION_MASK) !=
+         (iface->pkey & UCT_IB_PKEY_PARTITION_MASK))) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:588]: minor: (unpack_addr.pkey ^ iface->pkey) & UCT_IB_PKEY_PARTITION_MASK

File: src/uct/ib/base/ib_iface.c [hunk 19]
```diff
@@ -626,40 +659,39 @@ void uct_ib_iface_fill_ah_attr_from_addr(uct_ib_iface_t *iface,
...
+    int has_pkey_limited  = 0;
     uint16_t pkey_index, port_pkey, pkey;
 
-    if ((config->pkey_value != UCS_HEXUNITS_AUTO) &&
-        (config->pkey_value > UCT_IB_PKEY_PARTITION_MASK)) {
-        ucs_error("Requested pkey 0x%x is invalid, should be in the range 0..0x%x",
-                  config->pkey_value, UCT_IB_PKEY_PARTITION_MASK);
-        return UCS_ERR_INVALID_PARAM;
-    }
+    ucs_assert(!allow_limited || (config->pkey == UCS_HEXUNITS_AUTO));
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:694]: pls add comment

File: src/uct/ib/base/ib_iface.c [hunk 20]
```diff
@@ -673,43 +705,84 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
...
+                      pkey_index, pkey, UCT_IB_IFACE_ARG(iface));
+            has_pkey_limited = 1;
             continue;
         }
 
         /* take only the lower 15 bits for the comparison */
-        if ((config->pkey_value == UCS_HEXUNITS_AUTO) ||
-            ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey_value)) {
+        if ((config->pkey == UCS_HEXUNITS_AUTO) ||
+            ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:724]: seems the compare does not distinguish between limited and unlimited key?

File: src/uct/ib/base/ib_iface.c [hunk 21]
```diff
@@ -673,43 +705,84 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
...
 
     if (!pkey_found) {
-        if (config->pkey_value == UCS_HEXUNITS_AUTO) {
-            ucs_error("There is no valid pkey with full membership on "
-                      UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+        if (config->pkey == UCS_HEXUNITS_AUTO) {
+            if (!allow_limited && has_pkey_limited) {
+                ucs_debug("there is no valid pkey with full membership on "
+                          UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
+                return UCS_INPROGRESS;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:737]: should return UCS_ERR_NO_ELEM if not found

File: test/gtest/uct/ib/test_ib.cc (hunk 4 above)
Comment (yosefe) [reply test/gtest/uct/ib/test_ib.cc:12]: can we avoid calling create_connected_entities() if they are created from another code?

File: test/gtest/uct/ib/test_ib_pkey.cc [hunk 22]
```diff
@@ -7,10 +7,12 @@
 
 
 class test_uct_ib_pkey : public test_uct_ib_with_specific_port {
-public:
+protected:
     test_uct_ib_pkey() {
-        m_pkey_value = 0;
-        m_pkey_index = 0;
+        m_pkey[0]       = 0;
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_pkey.cc:12]: use constant INVALID_PKEY (also when checking the pkey table entry)

File: src/uct/ib/base/ib_iface.c (hunk 9 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c]: done

**Comment by dmitrygx** [comment]
```
RPM build errors:
error: File not found by glob: /scrap/azure/agent-04/AZP_WORKSPACE/2/s/build-test/_rpm/usr/lib64/ucx/libuct_cuda_gdrcopy.so.*
    File not found by glob: /scrap/azure/agent-04/AZP_WORKSPACE/2/s/build-test/_rpm/usr/lib64/ucx/libuct_cuda_gdrcopy.so.*
```

File: src/uct/ib/base/ib_iface.c [hunk 23]
```diff
@@ -577,21 +592,20 @@ int uct_ib_iface_is_reachable(const uct_iface_h tl_iface,
...
     const uct_ib_address_t *ib_addr = (const void*)dev_addr;
-    uct_ib_unpack_address_t unpack_addr;
+    uct_ib_address_pack_params_t params;
 
-    uct_ib_address_unpack(ib_addr, &unpack_addr);
+    uct_ib_address_unpack(ib_addr, &params);
 
     if (/* at least one PKEY has to be with full membership */
-        (((unpack_addr.pkey | iface->pkey) & UCT_IB_PKEY_MEMBERSHIP_MASK) == 0) ||
+        !(((params.pkey | iface->pkey) & UCT_IB_PKEY_MEMBERSHIP_MASK)) ||
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:609]: seems like one pair of ( ) can be removed

File: src/uct/ib/base/ib_iface.c (hunk 21 above)
Comment (yosefe) [reply src/uct/ib/base/ib_iface.c:767]: yes, but the error code from this function should reflect that this function did not find a pkey
the called may choose to continue if status==UCS_ERR_NO_ELEM, the further action based on the return status is outside the scope of this function

File: test/gtest/uct/ib/test_ib_pkey.cc [hunk 24]
```diff
@@ -182,7 +185,10 @@ UCS_TEST_P(test_uct_ib_pkey, test_pkey_pairs) {
         m_e2 = uct_test::create_entity(0);
         m_entities.push_back(m_e2);
 
-        prepare_entities();
+        m_e1->connect(0, *m_e2, 0);
+        m_e2->connect(0, *m_e1, 0);
+
+        prepare_entities(false);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib_pkey.cc:191]: prepare_entities(false) only does couple of EXPECT_EQ checks
maybe they could go to separate helper func, instead of passing true/false?

File: src/uct/ib/base/ib_iface.c [hunk 25]
```diff
@@ -431,11 +431,9 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
 void uct_ib_address_unpack(const uct_ib_address_t *ib_addr,
                            uct_ib_address_pack_params_t *params_p)
 {
-    const void *ptr = ib_addr + 1;
-    uct_ib_address_pack_params_t params;
+    const void *ptr                     = ib_addr + 1;
+    uct_ib_address_pack_params_t params = {0};
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:435]: add comment for cppcheck

Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:435]: done

File: src/uct/ib/base/ib_iface.c (hunk 23 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:609]: yes, removed

File: src/uct/ib/base/ib_iface.c (hunk 21 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:767]: yes, makes sense to me

File: test/gtest/uct/ib/test_ib_pkey.cc (hunk 24 above)
Comment (dmitrygx) [reply test/gtest/uct/ib/test_ib_pkey.cc:191]: yes, good catch
it would be better
done

**Comment by yosefe** [comment]
```
[2020-05-25T08:56:08.394Z] /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/base/ib_iface.c: In function ‘uct_ib_iface_init_pkey’:
[2020-05-25T08:56:08.394Z] /scrap/jenkins/workspace/ucx/contrib/../src/uct/ib/base/ib_iface.c:718: 
error: ‘lim_pkey_index’ may be used uninitialized in this function
```

File: src/uct/ib/base/ib_iface.c [hunk 26]
```diff
@@ -740,17 +742,19 @@ static ucs_status_t uct_ib_iface_search_pkey(uct_ib_iface_t *iface,
...
-        }
-
         /* take only the lower 15 bits for the comparison */
         if ((config->pkey == UCS_HEXUNITS_AUTO) ||
             ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
+            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) &&
+                (/* limited PKEY hasn't been find yet */
+                 (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) ||
+                 /* find more suitable PKEY with limited membership */
+                 ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey))) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:753]: can we unit this if with the if in lines 746-747? seems there is some duplication

File: test/gtest/uct/ib/test_ib.cc [hunk 27]
```diff
@@ -98,11 +96,14 @@ class test_uct_ib_addr : public test_uct_ib {
...
         pack_params.gid_index = std::numeric_limits<uint8_t>::max();
         pack_params.pkey      = iface->pkey;
+        address_size          = uct_ib_address_size(&params);
+        ib_addr               = (uct_ib_address_t*)malloc(address_size);
         uct_ib_address_pack(&pack_params, ib_addr);
 
         uct_ib_address_pack_params_t unpack_params;
         uct_ib_address_unpack(ib_addr, &unpack_params);
 
+
```
Comment (yosefe) [inline test/gtest/uct/ib/test_ib.cc:107]: why needed?

File: src/uct/ib/base/ib_iface.c (hunk 26 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:753]: reworked to make it simpler
could you check pls?

File: test/gtest/uct/ib/test_ib.cc (hunk 27 above)
Comment (dmitrygx) [reply test/gtest/uct/ib/test_ib.cc:107]: leftover

File: src/uct/ib/base/ib_iface.c [hunk 28]
```diff
@@ -743,23 +750,29 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
...
+        } else if (/* a specific PKEY was not requested and no
+                    * PKEY with limited membership was choosen */
+                   ((config->pkey == UCS_HEXUNITS_AUTO) &&
+                    (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY)) ||
+                   /* matches the requested PKEY */
+                   (uct_ib_iface_config_check_pkey(config, pkey))) {
+            /* save PKEY with limited membership for futher use if more suitable
+             * PKEY is not found */
+            lim_pkey_index = pkey_index;
+            lim_pkey       = pkey;
```
Comment (yosefe) [inline src/uct/ib/base/ib_iface.c:775]: this looks even more complicated..

Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c]: thanks for pointing me out at this code, now it fixed to be simple and w/o additional `if`s that were not needed

File: src/uct/ib/base/ib_iface.c [hunk 29]
```diff
@@ -354,30 +372,44 @@ void uct_ib_address_pack(const uct_ib_address_pack_params_t *params,
         ib_addr->flags |= UCT_IB_ADDRESS_FLAG_GID_INDEX;
         *(uint8_t*)ptr  = params->gid_index;
     }
+
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_PKEY) {
+        ucs_assert(params->pkey != UCT_IB_ADDRESS_DEFAULT_PKEY);
+        ib_addr->flags  |= UCT_IB_ADDRESS_FLAG_PKEY;
+        *(uint16_t*)ptr  = params->pkey;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:379]: extra space here and above

File: src/uct/ib/base/ib_iface.c [hunk 30]
```diff
@@ -386,95 +418,124 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
-        gid->global.subnet_prefix = UCT_IB_LINK_LOCAL_PREFIX; /* Default prefix */
-        gid->global.interface_id  = 0;
+        /* Default prefix */
+        params.gid.global.subnet_prefix = UCT_IB_LINK_LOCAL_PREFIX;
+        params.flags                   |= UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX;
 
-        /* If the link layer is not ETHERNET, then it is IB and a lid must be present */
-        *lid = *(const uint16_t*)ptr;
-        ptr  = UCS_PTR_TYPE_OFFSET(ptr, uint16_t);
+        params.gid.global.interface_id = 0;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:457]: alignment by `=`
also extra line above

File: src/uct/ib/base/ib_iface.c [hunk 31]
```diff
@@ -386,95 +418,124 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
+        params.flags                  |= UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:458]: minor can do assignment once, like:
```
 params.flags                  |= UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX |
                                            UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID;
```

File: src/uct/ib/base/ib_iface.c [hunk 32]
```diff
@@ -386,95 +418,124 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
+
+        /* If the link layer is not ETHERNET, then it is IB and a lid
+         * must be present */
+        params.lid = *(const uint16_t*)ptr;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:462]: i'd align with params filling above

File: src/uct/ib/base/ib_iface.c [hunk 33]
```diff
@@ -386,95 +418,124 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
+            params.flags                   |= UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX;
         }
     }
 
     if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_PATH_MTU) {
-        *path_mtu = *(const uint8_t*)ptr;
-        ptr       = UCS_PTR_TYPE_OFFSET(ptr, const uint8_t);
+        params.path_mtu = *(const uint8_t*)ptr;
+        ptr             = UCS_PTR_TYPE_OFFSET(ptr, const uint8_t);
+        params.flags    |= UCT_IB_ADDRESS_PACK_FLAG_PATH_MTU;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:487]: alignment

File: src/uct/ib/base/ib_iface.c [hunk 34]
```diff
@@ -386,95 +418,124 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
...
+        params.gid_index = *(const uint8_t*)ptr;
+        ptr              = UCS_PTR_TYPE_OFFSET(ptr, const uint16_t);
+        params.flags    |= UCT_IB_ADDRESS_PACK_FLAG_GID_INDEX;
+    }
+
+    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_PKEY) {
+        params.pkey = *(const uint16_t*)ptr;
     }
+    /* PKEY is always in params */
+    params.flags   |= UCT_IB_ADDRESS_PACK_FLAG_PKEY;
```
Comment (brminich) [inline src/uct/ib/base/ib_iface.c:500]: strange alignment

File: src/uct/ib/base/ib_iface.c (hunk 33 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:487]: done

File: src/uct/ib/base/ib_iface.c (hunk 34 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:500]: fixed

File: src/uct/ib/base/ib_iface.c (hunk 32 above)
Comment (dmitrygx) [reply src/uct/ib/base/ib_iface.c:462]: done

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by dmitrygx** [comment]
@yosefe ok to merge?
